core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -542,6 +552,8 @@ public ShareAcquiredRecords acquire(
         RecordBatch firstBatch = fetchPartitionData.records.batches().iterator().next();
         lock.writeLock().lock();
         try {
+            // Update the latest fetch offset metadata for any future queries.
+            this.latestFetchOffsetMetadata = fetchOffsetMetadata;
```

Comment:
The `acquire` call always comes after `DelayedShareFetch.tryComplete`, which already updates  `latestFetchOffsetMetadata`. So, it seems that we don't need to update `latestFetchOffsetMetadata` in `acquire`?

Reply:
`updateLatestFetchOffsetMetadata()` is only called from `tryComplete` when we find that there is a share partition whose offset metadata is not present (because in that case only, we call `readFromLog`). However, for any other case, we can only update the `latestFetchOffsetMetadata` via `acquire` method only. Else we'll have to call `readFromLog` from `tryComplete` in all scenarios to directly update share partition's `latestFetchOffsetMetadata`. Anything wrong in that understanding?

Reply:
Hmm, to me `acquire` seems to be the wrong place to update `latestFetchOffsetMetadata`. We update `latestFetchOffsetMetadata` with the fetch offset and then acquire a few batches after the fetch offset. This typically moves the next fetch offset, which makes `latestFetchOffsetMetadata` useless for the next fetch.

Reply:
That's my queustion with this comment https://github.com/apache/kafka/pull/17539#discussion_r1816703702. Do we ever get the lastFetchOffsetMetadata or we always update with fetchOffsetMetadata?

Reply:
Hi @junrao , I understand your point now. Here's my proposed solution changes- 
1. Remove update of `latestFetchOffsetMetadata` via `acquire` method
2. In `onComplete`, after I complete the call of `ShareFetchUtils.processFetchResponse`(which internally completes the call of `acquire` method), I will do a `readFromLog` for `topicPartitionData` with their latest fetch offset and update the share partition's `latestFetchOffsetMetadata`. Note this call to `readFromLog` will always happen despite the `tryComplete` `readFromLog` happens or not.
Please let me know what you think of this approach.